File: Stripe/STPAPIClient+PushProvisioning.m
//
//  STPAPIClient+PushProvisioning.m
//  Stripe
//
//  Created by Jack Flintermann on 9/27/18.
//  Copyright © 2018 Stripe, Inc. All rights reserved.
//

#import "STPAPIClient+PushProvisioning.h"
#import "STPAPIClient+Private.h"

@implementation STPAPIClient (PushProvisioning)

- (void)retrievePushProvisioningDetailsWithParams:(STPPushProvisioningDetailsParams *)params
                                     ephemeralKey:(STPEphemeralKey *)ephemeralKey
                                       completion:(STPPushProvisioningDetailsCompletionBlock)completion {
    
    NSString *endpoint = [NSString stringWithFormat:@"issuing/cards/%@/push_provisioning_details", params.cardId];
    NSDictionary *parameters = @{
                                 @"ios": @{
                                         @"certificates": params.certificatesBase64,
                                         @"nonce": params.nonceHex,
                                         @"nonce_signature": params.nonceSignatureHex,
                                         },
                                 };
    
    [STPAPIRequest<STPPushProvisioningDetails *> getWithAPIClient:self
                                                         endpoint:endpoint
                                                additionalHeaders:[self authorizationHeaderUsingEphemeralKey:ephemeralKey]
                                                       parameters:parameters
                                                     deserializer:[STPPushProvisioningDetails new]
                                                       completion:^(STPPushProvisioningDetails *details, __unused     NSHTTPURLResponse *response, NSError *error) {
        completion(details, error);
    }];
}
    
@end

void linkSTPAPIClientPushProvisioningCategory(void){}

File: Stripe/STPAPIClient+PushProvisioning.swift
//
//  STPAPIClient+PushProvisioning.swift
//  Stripe
//
//  Created by Jack Flintermann on 9/27/18.
//  Copyright © 2018 Stripe, Inc. All rights reserved.
//

import Foundation

typealias STPPushProvisioningDetailsCompletionBlock = (STPPushProvisioningDetails?, Error?) -> Void
extension STPAPIClient {
  @objc(retrievePushProvisioningDetailsWithParams:ephemeralKey:completion:)
  func retrievePushProvisioningDetails(
    with params: STPPushProvisioningDetailsParams,
    ephemeralKey: STPEphemeralKey,
    completion: @escaping STPPushProvisioningDetailsCompletionBlock
  ) {

    let endpoint = "issuing/cards/\(params.cardId)/push_provisioning_details"
    let parameters = [
      "ios": [
        "certificates": params.certificatesBase64,
        "nonce": params.nonceHex,
        "nonce_signature": params.nonceSignatureHex,
      ]
    ]

    APIRequest<STPPushProvisioningDetails>.getWith(
      self,
      endpoint: endpoint,
      additionalHeaders: authorizationHeader(using: ephemeralKey),
      parameters: parameters
    ) { details, _, error in
      completion(details, error)
    }
  }
}
